openfe/storage/warehouse.py

```diff
@@ -3,171 +3,95 @@
 
 
-class _DataContainer(abc.ABC):
+class WarehouseStores(TypedDict):
+    """This class serves to create a typesafe way of accessing the stores for
+    the WarehouseBaseClass.
     """
-    Abstract class, represents all data under some level of the heirarchy.
-    """
-
-    def __init__(self, parent, path_component):
-        self.parent = parent
-        self._path_component = self._to_path_component(path_component)
-        self._cache = {}
-
-    def __eq__(self, other):
-        return isinstance(other, self.__class__) and self.path == other.path
-
-    @staticmethod
-    def _to_path_component(item: Any) -> str:
-        """Convert input (object or string) to path string"""
-        if isinstance(item, str):
-            return item
-
-        # TODO: instead of str(hash(...)), this should return the digest
-        # that is being introduced in another PR; Python hash is not stable
-        # across sessions
-        return str(hash(item))
 
-    def __getitem__(self, item):
-        # code for the case this is a file
-        if item in self.external_storage:
-            return self.external_storage.load_stream(item)
+    setup: ExternalStorage
+    # We will add a result and task store here in the future.
 
-        # code for the case this is a "directory"
-        hash_item = self._to_path_component(item)
 
-        if hash_item not in self._cache:
-            self._cache[hash_item] = self._load_next_level(item)
+class WarehouseBaseClass:
+    def __init__(self, stores: WarehouseStores):
+        self.stores = stores
 
-        return self._cache[hash_item]
-
-    def __truediv__(self, item):
-        return self[item]
-
-    @abc.abstractmethod
-    def _load_next_level(self, item):
-        raise NotImplementedError()
-
-    def __iter__(self):
-        for loc in self.external_storage:
-            if loc.startswith(self.path):
-                yield loc
-
-    def load_stream(self, location):
-        return self.external_storage.load_stream(location)
-
-    def load_bytes(self, location):
-        with self.load_stream(location) as f:
-            byte_data = f.read()
-
-        return byte_data
-
-    @property
-    def path(self):
-        return self.parent.path + "/" + self._path_component
-
-    @property
-    def external_storage(self) -> ExternalStorage:
-        return self.parent.external_storage
+    def __eq__(self, other):
+        return isinstance(other, self.__class__) and self.stores == other.stores
 
     def __repr__(self):
         # probably should include repr of external store, too
-        return f"{self.__class__.__name__}({self.path})"
-
+        return f"{self.__class__.__name__}({self.stores})"
 
-class WarehouseBaseClass(_DataContainer):
-    def __init__(self, external_storage: ExternalStorage):
-        self._external_storage = external_storage
-        super().__init__(parent=self, path_component=None)
+    def delete(self, store_name: str, location: str):
```

Comment:
```suggestion
    def delete(self, store_name: Literal["setup"], location: str):
```

`store_name` should be a `Literal` yes? Also, it's unclear to me what `location` should be - it looks to be a gufe key where I see it used.

Reply:
That is correct. Much of this code is a bit strange because initially many parameters were untyped. I think that what we really want here is to explicitly state that we are accepting a `GufeKey` as the thing to be deleted. The problem is that as it exists, `delete` in `gufe.ExternalStorage` is actually accepting a `str`. My largest concern is that result objects might _not_ be of type `GufeKey`. For the time being, I am willing to change it and then update if we need to when we get there.